File: java/Samples/Biometrics/VB/ServerSampleVB/Controls/SettingsPanel.vb
Imports System
Imports System.Windows.Forms
Imports Neurotec.Biometrics

Partial Public Class SettingsPanel
	Inherits ControlBase
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()

		Dim thresholds() As Object = {Utilities.MatchingThresholdToString(12), Utilities.MatchingThresholdToString(24), Utilities.MatchingThresholdToString(36), Utilities.MatchingThresholdToString(48), Utilities.MatchingThresholdToString(60), Utilities.MatchingThresholdToString(72)}

		cbMatchingThreshold.Items.AddRange(thresholds)

		Dim speeds() As Object = {NMatchingSpeed.High, NMatchingSpeed.Medium, NMatchingSpeed.Low}
		cbFingersMatchingSpeed.Items.AddRange(speeds)
		cbFacesMatchingSpeed.Items.AddRange(speeds)
		cbIrisesMatchingSpeed.Items.AddRange(speeds)
		cbPalmsMatchingSpeed.Items.AddRange(speeds)
	End Sub

#End Region

#Region "Private nested types"

	Private Class Wrapper(Of T)
		Private privateValue As T
		Public Property Value() As T
			Get
				Return privateValue
			End Get
			Private Set(ByVal value As T)
				privateValue = value
			End Set
		End Property
		Private ReadOnly _displayName As String

		Public Sub New(ByVal value As T, ByVal displayName As String)
			value = value
			_displayName = displayName
		End Sub

		Public Overrides Function ToString() As String
			Return _displayName
		End Function

		Public Overloads Overrides Function Equals(ByVal obj As Object) As Boolean
			If TypeOf obj Is Wrapper(Of T) Then
				Dim objCast = TryCast(obj, Wrapper(Of T))
				Return objCast.Value.Equals(Value)
			End If
			If obj IsNot Nothing Then
				Return obj.Equals(Value)
			End If
			Return Value Is Nothing OrElse Value.Equals(obj)
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return Value.GetHashCode()
		End Function
	End Class

#End Region

#Region "Private form events"

	Private Sub MatchingThresholdValidating(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles cbMatchingThreshold.Validating
		Try
			Dim target = TryCast(sender, ComboBox)
			If target Is Nothing Then
				Return
			End If
			Dim value As Integer = Utilities.MatchingThresholdFromString(target.Text)
			target.Text = Utilities.MatchingThresholdToString(value)
		Catch
			e.Cancel = True
			MessageBox.Show("Matching threshold is invalid")
		End Try
	End Sub

	Private Sub BtnResetClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnReset.Click
		SettingsAccesor.ResetMatchingSettings()
		LoadSettings()
	End Sub

	Private Sub BtnApplyClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnApply.Click
		SettingsAccesor.MatchingThreshold = Utilities.MatchingThresholdFromString(cbMatchingThreshold.Text)
		SettingsAccesor.FingersMatchingSpeed = CType(cbFingersMatchingSpeed.SelectedItem, NMatchingSpeed)
		SettingsAccesor.FingersMaximalRotation = Utilities.MaximalRotationFromDegrees(Convert.ToInt32(nudFingersMaxRotation.Value))
		SettingsAccesor.FacesMatchingSpeed = CType(cbFacesMatchingSpeed.SelectedItem, NMatchingSpeed)
		SettingsAccesor.IrisesMatchingSpeed = CType(cbIrisesMatchingSpeed.SelectedItem, NMatchingSpeed)
		SettingsAccesor.IrisesMaximalRotation = Utilities.MaximalRotationFromDegrees(Convert.ToInt32(nudIrisesMaxRotation.Value))
		SettingsAccesor.PalmsMatchingSpeed = CType(cbPalmsMatchingSpeed.SelectedItem, NMatchingSpeed)
		SettingsAccesor.PalmsMaximalRotation = Utilities.MaximalRotationFromDegrees(Convert.ToInt32(nudPalmsMaxRotation.Value))
		SettingsAccesor.SaveSettings()
	End Sub

	Private Sub SettingsPanelLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		LoadSettings()
	End Sub

#End Region

#Region "Public methods"

	Public Overrides Function GetTitle() As String
		Return "Matching parameters"
	End Function

#End Region

#Region "Private methods"

	Private Sub LoadSettings()
		SelectThreshold(cbMatchingThreshold, SettingsAccesor.MatchingThreshold)
		cbFingersMatchingSpeed.SelectedItem = SettingsAccesor.FingersMatchingSpeed

		nudFingersMaxRotation.Value = Utilities.MaximalRotationToDegrees(SettingsAccesor.FingersMaximalRotation)
		nudIrisesMaxRotation.Value = Utilities.MaximalRotationToDegrees(SettingsAccesor.IrisesMaximalRotation)
		nudPalmsMaxRotation.Value = Utilities.MaximalRotationToDegrees(SettingsAccesor.PalmsMaximalRotation)
		cbFacesMatchingSpeed.SelectedItem = SettingsAccesor.FacesMatchingSpeed
		cbIrisesMatchingSpeed.SelectedItem = SettingsAccesor.IrisesMatchingSpeed
		cbPalmsMatchingSpeed.SelectedItem = SettingsAccesor.PalmsMatchingSpeed
	End Sub

	Private Sub SelectThreshold(ByVal target As ComboBox, ByVal value As Integer)
		Dim str As String = Utilities.MatchingThresholdToString(value)
		Dim index As Integer = target.Items.IndexOf(str)
		If index <> -1 Then
			target.SelectedIndex = index
		Else
			target.Items.Insert(0, str)
			target.SelectedIndex = 0
		End If
	End Sub

#End Region
End Class

File: java/Samples/Biometrics/CS/ServerSampleCS/Controls/SettingsPanel.cs
using System;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Samples.Code;

namespace Neurotec.Samples.Controls
{
	public partial class SettingsPanel : ControlBase
	{
		#region Public constructor

		public SettingsPanel()
		{
			InitializeComponent();

			object[] thresholds =
			{
				Utilities.MatchingThresholdToString(12),
				Utilities.MatchingThresholdToString(24),
				Utilities.MatchingThresholdToString(36),
				Utilities.MatchingThresholdToString(48),
				Utilities.MatchingThresholdToString(60),
				Utilities.MatchingThresholdToString(72)
			};

			cbMatchingThreshold.Items.AddRange(thresholds);

			object[] speeds = { NMatchingSpeed.High, NMatchingSpeed.Medium, NMatchingSpeed.Low };
			cbFingersMatchingSpeed.Items.AddRange(speeds);
			cbFacesMatchingSpeed.Items.AddRange(speeds);
			cbIrisesMatchingSpeed.Items.AddRange(speeds);
			cbPalmsMatchingSpeed.Items.AddRange(speeds);
		}

		#endregion

		#region Private nested types

		private class Wrapper<T>
		{
			public T Value { get; private set; }
			private readonly string _displayName;

			public Wrapper(T value, string displayName)
			{
				Value = value;
				_displayName = displayName;
			}

			public override string ToString()
			{
				return _displayName;
			}

			public override bool Equals(object obj)
			{
				if (obj is Wrapper<T>)
				{
					var objCast = obj as Wrapper<T>;
					return objCast.Value.Equals(Value);
				}
				if (obj != null)
				{
					return obj.Equals(Value);
				}
				return Value == null || Value.Equals(obj);
			}

			public override int GetHashCode()
			{
				return Value.GetHashCode();
			}
		}

		#endregion

		#region Private form events

		private void MatchingThresholdValidating(object sender, System.ComponentModel.CancelEventArgs e)
		{
			try
			{
				var target = sender as ComboBox;
				if (target == null) return;
				int value = Utilities.MatchingThresholdFromString(target.Text);
				target.Text = Utilities.MatchingThresholdToString(value);
			}
			catch
			{
				e.Cancel = true;
				MessageBox.Show(@"Matching threshold is invalid");
			}
		}

		private void BtnResetClick(object sender, EventArgs e)
		{
			SettingsAccesor.ResetMatchingSettings();
			LoadSettings();
		}

		private void BtnApplyClick(object sender, EventArgs e)
		{
			SettingsAccesor.MatchingThreshold = Utilities.MatchingThresholdFromString(cbMatchingThreshold.Text);
			SettingsAccesor.FingersMatchingSpeed = (NMatchingSpeed)cbFingersMatchingSpeed.SelectedItem;
			SettingsAccesor.FingersMaximalRotation = Utilities.MaximalRotationFromDegrees(Convert.ToInt32(nudFingersMaxRotation.Value));
			SettingsAccesor.FacesMatchingSpeed = (NMatchingSpeed)cbFacesMatchingSpeed.SelectedItem;
			SettingsAccesor.IrisesMatchingSpeed = (NMatchingSpeed)cbIrisesMatchingSpeed.SelectedItem;
			SettingsAccesor.IrisesMaximalRotation = Utilities.MaximalRotationFromDegrees(Convert.ToInt32(nudIrisesMaxRotation.Value));
			SettingsAccesor.PalmsMatchingSpeed = (NMatchingSpeed)cbPalmsMatchingSpeed.SelectedItem;
			SettingsAccesor.PalmsMaximalRotation = Utilities.MaximalRotationFromDegrees(Convert.ToInt32(nudPalmsMaxRotation.Value));
			SettingsAccesor.SaveSettings();
		}

		private void SettingsPanelLoad(object sender, EventArgs e)
		{
			LoadSettings();
		}

		#endregion

		#region Public methods

		public override string GetTitle()
		{
			return "Matching parameters";
		}

		#endregion

		#region Private methods

		private void LoadSettings()
		{
			SelectThreshold(cbMatchingThreshold, SettingsAccesor.MatchingThreshold);
			cbFingersMatchingSpeed.SelectedItem = SettingsAccesor.FingersMatchingSpeed;

			nudFingersMaxRotation.Value = Utilities.MaximalRotationToDegrees(SettingsAccesor.FingersMaximalRotation);
			nudIrisesMaxRotation.Value = Utilities.MaximalRotationToDegrees(SettingsAccesor.IrisesMaximalRotation);
			nudPalmsMaxRotation.Value = Utilities.MaximalRotationToDegrees(SettingsAccesor.PalmsMaximalRotation);
			cbFacesMatchingSpeed.SelectedItem = SettingsAccesor.FacesMatchingSpeed;
			cbIrisesMatchingSpeed.SelectedItem = SettingsAccesor.IrisesMatchingSpeed;
			cbPalmsMatchingSpeed.SelectedItem = SettingsAccesor.PalmsMatchingSpeed;
		}

		private void SelectThreshold(ComboBox target, int value)
		{
			string str = Utilities.MatchingThresholdToString(value);
			int index = target.Items.IndexOf(str);
			if (index != -1)
			{
				target.SelectedIndex = index;
			}
			else
			{
				target.Items.Insert(0, str);
				target.SelectedIndex = 0;
			}
		}

		#endregion
	}

}
